resolver/rewriter_resolver.go

```diff
@@ -70,13 +72,23 @@ func (r *RewriterResolver) Resolve(request *model.Request) (*model.Response, err
 	logger.WithField("resolver", Name(r.inner)).Trace("go to inner resolver")
 
 	response, err := r.inner.Resolve(request)
-	if err != nil {
-		return response, err
-	}
+	// Test for error after checking for fallbackOnFail
 
 	// Revert the request: must be done before calling r.next
 	request.Req = original
 
+	fallbackCondition := err != nil || (response != NoResponse && response.Res.Answer == nil)
```

Comment:
IMO querying the upstream when there's an error is not desirable, returning the error to the client seems more logical to me. A concrete case is we're going to leak private DNS requests when the server is down.  
If I understand your use case properly, falling back only when we get `NXDOMAIN` should be enough.  
Is that correct? What do you think about doing it that way?

It would also make the code simpler: the error handling would go back to the classic pattern of being handled immediately. And you could move the fallback block below the `NoResponse`.  
With both of those, the `fallbackCondition` would just be `response.Res.Answer == nil`.

Reply:
I had done that before, the problem is, my upstream for the internal domain times out slower than blocky. I do not want to increase blocky's timeout (unless there is a separate timeout for the conditional block) since that would slow down the resolution for other domains as well. Not just that, I would like a fallback when the resolver is down too.

My argument for your concern is that as soon as you opt for `fallbackUpstream`, you are forgoing the DNS leak protection. It actually makes sense because if you in any case want your DNS queries to go upstream for a domain, it means that the domain is already known to the upstream, just not particular subdomains. Remember that we are not forwarding the query after the mapping, we are forwarding the initial request itself. Meaning, public.site is the one that is sent out, not private.lan which is resolved by the internal resolver only.

In my usecase, I would _*prefer*_ that the queries be sent upstream when the internal resolver is down, so that I can access my services, albeit with a slower, through the internet, connection as opposed to a local one.

Reply:
I understand, I think it's fine as is then.

Maybe something to tackle as part of #476.